### ISSUE TITLE
Highlevel Workflow

### ISSUE THREAD

**Issue by ericxtang** [livepeer/livepeer-truebit-bigchaindb#1]
Current the Livepeer / Truebit integration works in the following way

![livepeer _ truebit integration 1](https://user-images.githubusercontent.com/345743/43529359-1adda9b0-9579-11e8-808d-9b60c270877b.png)

[source](https://sequencediagram.org/index.html#initialData=C4S2BsFMAIBkQG6QA6UgJ2gemgFXQK6QBGY0AkgHbCQDm6AhqAPaUBQbsACtAELrMGAEwDGDAM41MAWgB8cHvgaVxI5kIzQAXNABSzYgCVIARzbhmzZNABy0UAFtI4zosYq1GmfO553q9QwdJQ9AgAoAZTonagBKNkhKIVc-ZQCvaDkFaC4BYGY1cB0AYXAGEAcwuyjaGOBxeN9c5nzCzJ83NM8g6GKACwZwKEpaGBq66AAPFJD0zSzcXh0ANQwQADMQMRZKaGMTIklIuimAGlSVYtZN2njF9uzm1uYi6FX0Da2mEFY95wJwMAgA)

Note when `LP Transcoder` submits `Verification Request` to `TB`, the segment information is submitted.

Currently the segment information is represented by its IPFS hash, which means `LP Transcoder` has to operate an IPFS node, and `TB` has to interface with IPFS to load the data. There is no guarantee the IPFS hash submitted will be valid.  
_Note `LP Transcoder` is incentivized to submit the correct data_

If `TB` has integration with BigchainDB, that would be another way for `LP Transcoder` to transfer the data.

Let's discuss the pros and cons of this approach.

**Comment by OR13** [comment]
For sure, one pro of this approach is to abstract the data availability problem into an adapter interface. 

I think it would be best to support arbitrary data providers, not everyone will want to use IPFS or BigchainDB. 

Using an adapter will force the design to respect separation of concerns, make testability easier and decoupled from storage, etc.

Seems safe to keep the interface key, value, content addressed, I love IPFS, but it isn't the fastest or only way to do this.

**Comment by ericxtang** [comment]
@OR13 adapters seem like a good option for an implementation like this.

It's interesting to think about the tradeoffs and the different advantages offered by the storage solutions.  For example, our hope here with bigchainDB is to create a POA-style datastore to solve the problem where IPFS cannot currently guarantee the delivery of a specific piece of data. This is an attack vector to trigger a slashing condition on an honest actor.

**Comment by OR13** [comment]
Seems like we should be clearer on what a valid segment identifier is. Currently I would suggest that it is an IPFS hash which is available from some IPFS server with pinning turned on (hosted by TrueBit or Liverpeer). For the hash to be valid it should be resolvable to binary within the time window to avoid slashing.

To generalize this we need to know, how long is the max time a client can wait for a hash to resolve (before slashing is triggered).

Then I would argue the interface for the adapter could be as follows:

```
async get_data(trusted_host_rest_api, data_hash) => binary_data
async set_data(trusted_host_rest_api, data) => data_hash
```
In the case of IPFS, the host should be a an IPFS node with pinning enabled, that is trusted by both livepeer and truebit clients.

Other than the timing requirements, what other details are required for this interface to be fully specified?

**Comment by yondonfu** [comment]
>  Currently I would suggest that it is an IPFS hash which is available from some IPFS server with pinning turned on

In the current implementation of the Livepeer protocol, when a transcoder submits for verification it provides a `dataStorageHash` (named as such to make it clear that it is a content addressed identifier as opposed to just computing the keccak256 hash of the data) which is the IPFS hash identifying the stored segment data which the transcoder uploads to IPFS (with pinning turned on).

Would the adapter interface expect the same identifier scheme to be used for all underlying implementations? I ask because IPFS hashes are generated using a specific chunking algorithm and I'm not sure what the identifiers used for BigChainDB are.

**Comment by OR13** [comment]
Good question, assuming it will eventually be bytes32, it might make sense to use bytes32 as the identifier in BigChainDB, and convert IPFS hash to that... IPFS multihashes are too big, we have used this code for a similar adapter for javascript objects (feel free to take it and strip the JSON bits):

https://github.com/transmute-industries/transmute/blob/master/packages/transmute-framework/src/decentralized-storage/ipfs/index.js#L80

You can also store the multihash without stripping the first 2 bytes, using this approach:

https://ethereum.stackexchange.com/questions/17094/how-to-store-ipfs-hash-using-bytes

https://github.com/ipfs/faq/issues/22

TLDR, multihashes are too big to store in bytes32, either you store a struct representing the multihash or you omit the hashing algorithm (assume sha-256, the first approach). I recommend the latter, a simple hex content identifier (sha-256-256(data)) be used as the key for all storage adapters.
